drivers/sensor/si1153/si1153.c

```diff
@@ -0,0 +1,663 @@
+				 *  overflow them all so they all have the same reference
+				 */
+				ps_exit_time[0] += 0x1FFFFFFFL;
+				ps_exit_time[1] += 0x1FFFFFFFL;
+				ps_exit_time[2] += 0x1FFFFFFFL;
+
+				ps_entry_time[0] += 0x1FFFFFFFL;
+				ps_entry_time[1] += 0x1FFFFFFFL;
+				ps_entry_time[2] += 0x1FFFFFFFL;
+			}
+
+			/**
+			 * Calculate the midpoint (between entry and exit times)
+			 * of each waveform the order of these midpoints helps
+			 * determine the gesture
+			 */
+			ps_time[0] = CALCULATE_MID_POINT(ps_entry_time[0], ps_exit_time[0]);
+			ps_time[1] = CALCULATE_MID_POINT(ps_entry_time[1], ps_exit_time[1]);
+			ps_time[2] = CALCULATE_MID_POINT(ps_entry_time[2], ps_exit_time[2]);
+
+			/**
+			 * The diff_x and diff_y values help determine a gesture by
+			 * comparing the LED measurements that are on a single axis
+			 */
+			diff_x  = CALCULATE_ABS_DIFF(ps_time[1], ps_time[2]);
+			diff_y1 = CALCULATE_ABS_DIFF(ps_time[0], ps_time[1]);
+			diff_y2 = CALCULATE_ABS_DIFF(ps_time[0], ps_time[2]);
+
+			/**
+			 * Take the average of all three midpoints to make
+			 * a comparison point for each midpoint
+			 */
+			if ((ps_exit_time[0] - ps_entry_time[0]) > GESTURE_TIME_DIFF ||
+				(ps_exit_time[1] - ps_entry_time[1]) > GESTURE_TIME_DIFF ||
+				(ps_exit_time[2] - ps_entry_time[2]) > GESTURE_TIME_DIFF) {
+				if( ( (ps_time[0] < ps_time[1]) &&  (diff_y1 > diff_x) ) ||
+					( (ps_time[0] <= ps_time[2]) && (diff_y2 > diff_x) ) ){
+					/**
+					 * An up gesture occured if the bottom LED had its midpoint first
+					 */
+					ret = UP;
+				}
+				else if ( ( (ps_time[0] < ps_time[1]) && (diff_y1 > diff_x) ) ||
+						  ( (ps_time[0] > ps_time[2]) && (diff_y2 > diff_x) ) ) {
+					/**
+					 * A down gesture occured if the bottom LED had its midpoint last
+					 */
+					ret = DOWN;
+				}
+				else if((ps_time[0] < ps_time[1]) &&
+						(ps_time[2] < ps_time[1]) &&
+						(diff_x > ((diff_y1+diff_y2)/2))) {
+					/**
+					 * A left gesture occured if the left LED had its midpoint last
+					 */
+					ret = LEFT;
+				}
+				else if( (ps_time[0] < ps_time[2]) &&
+						(ps_time[1] < ps_time[2]) &&
+						(diff_x > ((diff_y1+diff_y2)/2))) {
+					/**
+					 * A right gesture occured if the right LED had midpoint later than the right LED
+					 */
+					ret = RIGHT;
+				}
+			}
+		}
+
+		for (i = 0; i < 3; i++) {
+			ps_exit_time[i]  = 0;
+			ps_entry_time[i] = 0;
+		}
+
+	}
+
+	return ret;
+}
+
+/**
+ * Configuration for left, right and bottom channel.
+ */
+#define ADCCONFIG   (ADCCFG_DR_4096 | ADCCFG_AM_LARGE_IR)
+#define ADCSENS     (ADCSENS_HSIG_NORM | ADCSENS_SW_GAIN_4MEAS | ADCSENS_HW_GAIN_48_8US)
+#define ADCPOST     (ADCPOST_24BIT | ADCPOST_POSTSHIFT_0 | ADCPOST_THRESH_EN_0)
+#define MEASURECFG  (MEASCFG_NO_MEAS | MEASCFG_LED_NOM | MEASCFG_BANK_SEL_A)
+
+static int si115x_init(struct si1153_data *data)
+{
+	int retval;
+
+	retval = si115x_reset(data);
+
+	retval += si115x_param_set(data, PARAM_CH_LIST, 0x0f);
```

Comment:
I'd use |= instead of += to avoid overflows that could result retval adding up to 0.